### ISSUE TITLE
Infinite loop when logging out

### ISSUE THREAD

**Issue by kkfrosty** [IdentityServer/IdentityServer3#1814]
The scenario is we had to come up with a way determine if a user was an internal windows user and utilize our custom user database to manually authenticate them if the windows account matches an account in our user table.

The process for logging in we finally got to work.  We use the Thinktecture token client and call RequestResourceOwnerPasswordAsync.

We then we use the JwtSecurityTokenHandler ValidateToken() to get a ClaimsPrincipal.
We use our usermanager CreateIdentity to create our ClaimsIdentity.

We use similar code found in our AuthorizationCodeReceived delegate to extract claims.  The only thing we don't have available is an id_token.

However, from debugging it appears the access_token is the same as an IdToken.  So we have tried using that.

The problem is in the RedirectToIdentityProvider on logout.  IE goes into an infinite loop.  Looking at the Identity Server log, we see the following over and over.

w3wp.exe Information: 0 : [Thinktecture.IdentityServer.Core.Validation.EndSessionRequestValidator]: 9/1/2015 5:58:56 PM +00:00 -- Start end session request validation
w3wp.exe Warning: 0 : [Thinktecture.IdentityServer.Core.Validation.EndSessionRequestValidator]: 9/1/2015 5:58:56 PM +00:00 -- User is anonymous. Ignoring end session parameters
w3wp.exe Information: 0 : [Thinktecture.IdentityServer.Core.Results.LogoutResult]: 9/1/2015 5:58:56 PM +00:00 -- Redirecting to logout page
w3wp.exe Information: 0 : [Thinktecture.IdentityServer.Core.Endpoints.AuthenticationController]: 9/1/2015 5:58:57 PM +00:00 -- Logout endpoint submitted
w3wp.exe Information: 0 : [Thinktecture.IdentityServer.Core.Endpoints.AuthenticationController]: 9/1/2015 5:58:57 PM +00:00 -- Clearing cookies
w3wp.exe Information: 0 : [Thinktecture.IdentityServer.Core.Endpoints.AuthenticationController]: 9/1/2015 5:58:57 PM +00:00 -- rendering logged out page
w3wp.exe Information: 0 : [Thinktecture.IdentityServer.Core.Endpoints.AuthenticationController]: 9/1/2015 5:58:57 PM +00:00 -- Logout endpoint submitted
w3wp.exe Information: 0 : [Thinktecture.IdentityServer.Core.Endpoints.AuthenticationController]: 9/1/2015 5:58:57 PM +00:00 -- Clearing cookies
w3wp.exe Information: 0 : [Thinktecture.IdentityServer.Core.Endpoints.AuthenticationController]: 9/1/2015 5:58:57 PM +00:00 -- rendering logged out page
w3wp.exe Information: 0 : [Thinktecture.IdentityServer.Core.Endpoints.AuthenticationController]: 9/1/2015 5:58:57 PM +00:00 -- Logout endpoint submitted
w3wp.exe Information: 0 : [Thinktecture.IdentityServer.Core.Endpoints.AuthenticationController]: 9/1/2015 5:58:57 PM +00:00 -- Clearing cookies
w3wp.exe Information: 0 : [Thinktecture.IdentityServer.Core.Endpoints.AuthenticationController]: 9/1/2015 5:58:57 PM +00:00 -- rendering logged out page
w3wp.exe Information: 0 : [Thinktecture.IdentityServer.Core.Endpoints.AuthenticationController]: 9/1/2015 5:58:57 PM +00:00 -- Logout endpoint submitted
w3wp.exe Information: 0 : [Thinktecture.IdentityServer.Core.Endpoints.AuthenticationController]: 9/1/2015 5:58:57 PM +00:00 -- Clearing cookies
w3wp.exe Information: 0 : [Thinktecture.IdentityServer.Core.Endpoints.AuthenticationController]: 9/1/2015 5:58:57 PM +00:00 -- rendering logged out page
w3wp.exe Information: 0 : [Thinktecture.IdentityServer.Core.Endpoints.AuthenticationController]: 9/1/2015 5:58:57 PM +00:00 -- Logout endpoint submitted
w3wp.exe Information: 0 : [Thinktecture.IdentityServer.Core.Endpoints.AuthenticationController]: 9/1/2015 5:58:57 PM +00:00 -- Clearing cookies
w3wp.exe Information: 0 : [Thinktecture.IdentityServer.Core.Endpoints.AuthenticationController]: 9/1/2015 5:58:57 PM +00:00 -- rendering logged out page
w3wp.exe Information: 0 : [Thinktecture.IdentityServer.Core.Endpoints.AuthenticationController]: 9/1/2015 5:58:57 PM +00:00 -- Logout endpoint submitted
w3wp.exe Information: 0 : [Thinktecture.IdentityServer.Core.Endpoints.AuthenticationController]: 9/1/2015 5:58:57 PM +00:00 -- Clearing cookies
w3wp.exe Information: 0 : [Thinktecture.IdentityServer.Core.Endpoints.AuthenticationController]: 9/1/2015 5:58:57 PM +00:00 -- rendering logged out page
w3wp.exe Information: 0 : [Thinktecture.IdentityServer.Core.Endpoints.AuthenticationController]: 9/1/2015 5:58:57 PM +00:00 -- Logout endpoint submitted
w3wp.exe Information: 0 : [Thinktecture.IdentityServer.Core.Endpoints.AuthenticationController]: 9/1/2015 5:58:57 PM +00:00 -- Clearing cookies
w3wp.exe Information: 0 : [Thinktecture.IdentityServer.Core.Endpoints.AuthenticationController]: 9/1/2015 5:58:57 PM +00:00 -- rendering logged out page
w3wp.exe Information: 0 : [Thinktecture.IdentityServer.Core.Endpoints.AuthenticationController]: 9/1/2015 5:58:57 PM +00:00 -- Logout endpoint submitted
w3wp.exe Information: 0 : [Thinktecture.IdentityServer.Core.Endpoints.AuthenticationController]: 9/1/2015 5:58:57 PM +00:00 -- Clearing cookies
w3wp.exe Information: 0 : [Thinktecture.IdentityServer.Core.Endpoints.AuthenticationController]: 9/1/2015 5:58:57 PM +00:00 -- rendering logged out page
w3wp.exe Information: 0 : [Thinktecture.IdentityServer.Core.Endpoints.AuthenticationController]: 9/1/2015 5:58:57 PM +00:00 -- Logout endpoint submitted
w3wp.exe Information: 0 : [Thinktecture.IdentityServer.Core.Endpoints.AuthenticationController]: 9/1/2015 5:58:57 PM +00:00 -- Clearing cookies
w3wp.exe Information: 0 : [Thinktecture.IdentityServer.Core.Endpoints.AuthenticationController]: 9/1/2015 5:58:57 PM +00:00 -- rendering logged out page
w3wp.exe Information: 0 : [Thinktecture.IdentityServer.Core.Endpoints.AuthenticationController]: 9/1/2015 5:58:57 PM +00:00 -- Logout endpoint submitted
w3wp.exe Information: 0 : [Thinktecture.IdentityServer.Core.Endpoints.AuthenticationController]: 9/1/2015 5:58:57 PM +00:00 -- Clearing cookies
w3wp.exe Information: 0 : [Thinktecture.IdentityServer.Core.Endpoints.AuthenticationController]: 9/1/2015 5:58:57 PM +00:00 -- rendering logged out page

Any ideas of why code would be repeating this?

**Comment by kkfrosty** [comment]
Digging around more, since we're having to log the users in via code, it appears there will be no IdToken.

All we want the users to do is redirected back to the login page.  Tee infinite loop starts when the browser redirects to a logout page.

**Comment by brockallen** [comment]
If you're not using IdSvr for SSO, why do you want to send the users to its logout page? They have no login with IdSvr.

**Comment by kkfrosty** [comment]
Thanks you Brock.

We are using Identity Server but in this certain case we have to authenticate windows users so they don't have to type in a user name and password.
The system still has to work for anybody not in a domain.

 What we're doing is custom code because companies which run the app internally, they do not want their windows users to enter in a user name or password.  I WOULD LOVE another way than what we're doing.  However, we've been unable to find code to simulate what happens when a user is redirected to the login page and supplies credentials.

If the user is a windows user and is authenticated then we know what the user name and password are behind the scenes.  The best we could come up with and that would work reliably is what I posted above.

Ideally once they have a token, the system performs the same as if we logged them in or if they manually logged in from the login page.

So we don't have conditions in our start.auth.cs, so are you saying we should and not let the app.UseOpenIdConnectAuthentication code run?

While troubleshooting, I've found the code which appears to be kicking off the RedirectToIdentityProvider delegate.

Request.GetOwinContext().Authentication.SignOut();

For now it seems if I change that line of code to this, 
Request.GetOwinContext().Authentication.SignOut(DefaultAuthenticationTypes.ExternalCookie);

The problem appears to go away but then manual signin users don't get signed out.

So I created a condition based on the Id Token being present and this appears to work.  

        public ActionResult Signout()
        {
            // also possible to pass post logout redirect url via properties
            var properties = new AuthenticationProperties
            {
                RedirectUri = <OurRedirectURL>
            };

            if (Request.GetOwinContext().Authentication.User.HasClaim(c => c.Type == "id_token"))
            {
                Request.GetOwinContext().Authentication.SignOut();
            }
            else
            {
                Request.GetOwinContext().Authentication.SignOut(DefaultAuthenticationTypes.ExternalCookie);
            }
            

            //AuthenticationManager.SignOut(DefaultAuthenticationTypes.ExternalCookie);

           // return Redirect("/");

            return Redirect(<OurRedirectURL>]);
        }
